### PR TITLE
LibreSSL: 3.5.3 → 3.5.4, 3.6.1 → 3.6.2

### PR DESCRIPTION
**Pull request by ruuda**
###### Description of changes

Update both supported versions of LibreSSL to the latest patch release. Both were released February 7th and include an important security fix:

> A malicious certificate revocation list or timestamp response token would allow an attacker to read arbitrary memory.

* [3.6.2 release notes](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.6.2-relnotes.txt)
* [3.5.4 release notes](https://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-3.5.4-relnotes.txt)

I suspect that 3.4, which is still packaged, is also affected, but 3.4 was introduced in [OpenBSD 7.0](https://www.openbsd.org/70.html), which was released October 14th 2021, and LibreSSL branches are only maintained for one year after the corresponding OpenBSD release, so 3.4 is no longer supported.

<del>I will make a separate pull request to try and remove 3.4.</del><ins>There are two blockers for removing 3.4, I opened #216207 to track those.</ins>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by NickCao** [comment]
Result of `nixpkgs-review pr 216202` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>linuxKernel.packages.hardkernel_4_14.prl-tools</li>
    <li>linuxPackages_hardkernel_latest.prl-tools</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>linuxPackages_testing_bcachefs.prl-tools (linuxKernel.packages.linux_testing_bcachefs.prl-tools)</li>
  </ul>
</details>
<details>
  <summary>49 packages built:</summary>
  <ul>
    <li>acme-client</li>
    <li>catgirl</li>
    <li>citra-canary</li>
    <li>citra-nightly</li>
    <li>compactor</li>
    <li>disnix</li>
    <li>disnixos</li>
    <li>dydisnix</li>
    <li>dysnomia</li>
    <li>geomyidae</li>
    <li>gmid</li>
    <li>hydra_unstable</li>
    <li>kamid</li>
    <li>libressl (libressl_3_6)</li>
    <li>libressl_3_5</li>
    <li>linuxKernel.packages.linux_4_14.prl-tools</li>
    <li>linuxPackages_4_14_hardened.prl-tools (linuxKernel.packages.linux_4_14_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_4_19.prl-tools</li>
    <li>linuxPackages_4_19_hardened.prl-tools (linuxKernel.packages.linux_4_19_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_5_10.prl-tools</li>
    <li>linuxPackages_5_10_hardened.prl-tools (linuxKernel.packages.linux_5_10_hardened.prl-tools)</li>
    <li>linuxPackages.prl-tools (linuxKernel.packages.linux_5_15.prl-tools)</li>
    <li>linuxPackages_hardened.prl-tools (linuxPackages_5_15_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_5_4.prl-tools</li>
    <li>linuxPackages_latest.prl-tools (linuxKernel.packages.linux_6_1.prl-tools)</li>
    <li>linuxPackages_6_1_hardened.prl-tools (linuxKernel.packages.linux_6_1_hardened.prl-tools)</li>
    <li>linuxPackages_latest-libre.prl-tools (linuxKernel.packages.linux_latest_libre.prl-tools)</li>
    <li>linuxPackages-libre.prl-tools (linuxKernel.packages.linux_libre.prl-tools)</li>
    <li>linuxPackages_lqx.prl-tools (linuxKernel.packages.linux_lqx.prl-tools)</li>
    <li>linuxPackages_xanmod.prl-tools (linuxKernel.packages.linux_xanmod.prl-tools)</li>
    <li>linuxPackages_xanmod_latest.prl-tools (linuxKernel.packages.linux_xanmod_latest.prl-tools ,linuxPackages_xanmod_stable.prl-tools)</li>
    <li>linuxPackages_zen.prl-tools (linuxKernel.packages.linux_zen.prl-tools)</li>
    <li>lispPackages.nyxt</li>
    <li>mydumper</li>
    <li>netcat</li>
    <li>notemap</li>
    <li>nyxt</li>
    <li>openntpd (openntpd_nixos)</li>
    <li>opensmtpd</li>
    <li>perseus-cli</li>
    <li>pounce</li>
    <li>restream</li>
    <li>sacc</li>
    <li>telescope</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tinystatus</li>
    <li>twa</li>
    <li>wasm-pack</li>
    <li>waynergy</li>
  </ul>
</details>

**Review by NickCao — COMMENTED** [review]
`linuxPackages_testing_bcachefs.prl-tools` failure looks unrelated.

**Comment by NickCao** [comment]
@ofborg build opensmtpd acme-client telescope wasm-pack

**Comment by ajs124** [comment]
Feel free to add `knownVulnerabilities` to `libressl_3_4`. I forgot to follow up on https://github.com/NixOS/nixpkgs/issues/172151, otherwise it would already be gone.

**Comment by ruuda** [comment]
> Feel free to add knownVulnerabilities to libressl_3_4

TIL about `knownVulnerabilities`, thanks for pointing this out, I added it.

@botovq pointed out in #216207 that the fix for the vulnerability that triggered the release is small and we might backport it, I’ll try that as well.

File: pkgs/development/libraries/libressl/default.nix [hunk 1]
```diff
@@ -80,18 +86,34 @@ let
       license = with licenses; [ publicDomain bsdOriginal bsd0 bsd3 gpl3 isc openssl ];
       platforms   = platforms.all;
       maintainers = with maintainers; [ thoughtpolice fpletz ];
+      knownVulnerabilities = knownVulnerabilities;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libressl/default.nix:89]:
```suggestion
      inherit knownVulnerabilities;
```

File: pkgs/development/libraries/libressl/default.nix [hunk 2]
```diff
@@ -80,18 +86,34 @@ let
...
     version = "3.4.3";
     hash = "sha256-/4i//jVIGLPM9UXjyv5FTFAxx6dyFwdPUzJx1jw38I0=";
+    knownVulnerabilities = [
+      "Support ended 2022-10-14."
+      # The following vulnerability is originally present,
+      # but we backport the fix below.
+      # "https://marc.info/?l=libressl&m=167582148932407&w=2"
+    ];
+    patches = [
+      (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/libressl/default.nix:104]:
```suggestion
    ];
    patches = [
      # https://marc.info/?l=libressl&m=167582148932407&w=2
      (fetchpatch {
```

**Comment by ajs124** [comment]
what depends on `libressl_3_4` again? can we maybe just build those things against something supported?

IMO we should mark it as vulnerable still. As for the backport… I don't like breaking things on stable, but shipping likely insecure software also isn't great.

**Comment by ruuda** [comment]
> what depends on libressl_3_4 again? can we maybe just build those things against something supported?

It’s not that easy, I opened #216207 for it. The main remaining blocker at this point are all currently packaged versions of `foundationdb`, and they are incompatible with newer LibreSSL. [There is a work in progress to package a newer version of FoundationDB and build it against OpenSSL](https://github.com/NixOS/nixpkgs/pull/215336#issuecomment-1428535512), but I don’t know if the older versions can just be dropped once that is complete.

**Comment by ajs124** [comment]
Result of `nixpkgs-review pr 216202` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages marked as broken and skipped:</summary>
  <ul>
    <li>foundationdb</li>
    <li>foundationdb51</li>
    <li>foundationdb52</li>
    <li>foundationdb60</li>
    <li>foundationdb61</li>
    <li>libressl_3_4</li>
    <li>linuxKernel.packages.hardkernel_4_14.prl-tools</li>
    <li>linuxPackages_hardkernel_latest.prl-tools</li>
    <li>python310Packages.foundationdb51</li>
    <li>python310Packages.foundationdb52</li>
    <li>python310Packages.foundationdb60</li>
    <li>python310Packages.foundationdb61</li>
    <li>python311Packages.foundationdb51</li>
    <li>python311Packages.foundationdb52</li>
    <li>python311Packages.foundationdb60</li>
    <li>python311Packages.foundationdb61</li>
  </ul>
</details>
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>linuxPackages_testing_bcachefs.prl-tools (linuxKernel.packages.linux_testing_bcachefs.prl-tools)</li>
  </ul>
</details>
<details>
  <summary>51 packages built:</summary>
  <ul>
    <li>acme-client</li>
    <li>catgirl</li>
    <li>citra-canary</li>
    <li>citra-nightly</li>
    <li>compactor</li>
    <li>disnix</li>
    <li>disnixos</li>
    <li>dydisnix</li>
    <li>dysnomia</li>
    <li>geomyidae</li>
    <li>gmid</li>
    <li>hydra_unstable</li>
    <li>kamid</li>
    <li>libressl (libressl_3_6)</li>
    <li>libressl_3_5</li>
    <li>linuxKernel.packages.linux_4_14.prl-tools</li>
    <li>linuxPackages_4_14_hardened.prl-tools (linuxKernel.packages.linux_4_14_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_4_19.prl-tools</li>
    <li>linuxPackages_4_19_hardened.prl-tools (linuxKernel.packages.linux_4_19_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_5_10.prl-tools</li>
    <li>linuxPackages_5_10_hardened.prl-tools (linuxKernel.packages.linux_5_10_hardened.prl-tools)</li>
    <li>linuxPackages.prl-tools (linuxKernel.packages.linux_5_15.prl-tools)</li>
    <li>linuxPackages_hardened.prl-tools (linuxPackages_5_15_hardened.prl-tools)</li>
    <li>linuxKernel.packages.linux_5_4.prl-tools</li>
    <li>linuxKernel.packages.linux_6_1.prl-tools</li>
    <li>linuxPackages_6_1_hardened.prl-tools (linuxKernel.packages.linux_6_1_hardened.prl-tools)</li>
    <li>linuxPackages_latest-libre.prl-tools (linuxKernel.packages.linux_latest_libre.prl-tools)</li>
    <li>linuxPackages-libre.prl-tools (linuxKernel.packages.linux_libre.prl-tools)</li>
    <li>linuxPackages_lqx.prl-tools (linuxKernel.packages.linux_lqx.prl-tools)</li>
    <li>linuxPackages_xanmod.prl-tools (linuxKernel.packages.linux_xanmod.prl-tools)</li>
    <li>linuxPackages_xanmod_latest.prl-tools (linuxKernel.packages.linux_xanmod_latest.prl-tools ,linuxPackages_xanmod_stable.prl-tools)</li>
    <li>linuxPackages_zen.prl-tools (linuxKernel.packages.linux_zen.prl-tools)</li>
    <li>linuxPackages_latest.prl-tools</li>
    <li>lispPackages.nyxt</li>
    <li>mydumper</li>
    <li>netcat</li>
    <li>notemap</li>
    <li>nyxt</li>
    <li>openntpd (openntpd_nixos)</li>
    <li>opensmtpd</li>
    <li>ovftool</li>
    <li>perseus-cli</li>
    <li>pounce</li>
    <li>restream</li>
    <li>sacc</li>
    <li>telescope</li>
    <li>tests.testers.nixosTest-example</li>
    <li>tinystatus</li>
    <li>twa</li>
    <li>wasm-pack</li>
    <li>waynergy</li>
  </ul>
</details>

**Comment by ajs124** [comment]
Thanks for taking care of this. Can you open a backport with the three commits that aren't cb8d827b9a4c1863c37c0b124f78da93b3b5cb05?

AFAIK we don't really have any good policies around this stuff, so let's just go with best effort to ship secure software while not breaking the stable release, which in this case is backporting the fixes, but not the `knownVulnerabilities`.

**Comment by ruuda** [comment]
Makes sense! I opened #217749 for the backport.

**Pull request merged**